canonicalwebteam/blog/common_view_logic.py

```diff
@@ -5,39 +5,104 @@
+                upcoming, _ = api.get_articles(
+                    tags=self.tag_ids,
+                    tags_exclude=self.excluded_tags,
+                    page=page,
+                    per_page=3,
+                    categories=[events["id"], webinars["id"]],
+                )
+
+        articles, total_pages = api.get_articles(
+            tags=self.tag_ids,
+            tags_exclude=self.excluded_tags,
+            exclude=featured_article_ids,
+            page=page,
+            categories=[category_id],
+        )
+
+        context = get_index_context(
+            page,
+            articles,
+            total_pages,
+            featured_articles=featured_articles,
+            upcoming=upcoming,
+        )
+
+        context["title"] = self.blog_title
+        context["category"] = {"slug": category}
+        context["upcoming"] = upcoming
+
+        return context
+
+
 def get_complete_article(article, group=None):
     """
     This returns any given article from the wordpress API
     as an object that includes all information for the templates,
     some of which will be fetched from the Wordpress API
     """
-    featured_image = api.get_media(article["featured_media"])
+    featured_image = {}
+    if "wp:featuredmedia" in article["_embedded"]:
+        featured_image = article["_embedded"]["wp:featuredmedia"][0]
 
-    author = api.get_user(article["author"])
+    author = {}
+    if "author" in article["_embedded"]:
+        author = article["_embedded"]["author"][0]
+
+    categories = []
+    if "wp:term" in article["_embedded"]:
+        categories = article["_embedded"]["wp:term"][0]
 
-    category_ids = article["categories"]
+    for category in categories:
+        if category["id"] not in category_cache:
+            category_cache[category["id"]] = category
 
-    # Can these calls be bundled?
-    first_item = True
-    for category_id in category_ids:
-        if category_id not in category_cache:
-            resolved_category = api.get_category_by_id(category_id)
-            category_cache[category_id] = resolved_category
-        if first_item:
-            article["display_category"] = category_cache[category_id]
-            first_item = False
+        if "display_category" not in article:
+            article["display_category"] = category
 
     if group:
         article["group"] = group
     else:
-        first_item = True
-        for group_id in article["group"]:
-            if group_id not in group_cache:
-                resolved_group = api.get_group_by_id(group_id)
-                group_cache[group_id] = resolved_group
-            if first_item:
-                article["group"] = group_cache[group_id]
-                first_item = False
+        if (
+            "wp:term" in article["_embedded"]
+            and article["_embedded"]["wp:term"][3]
+        ):
+            article["group"] = article["_embedded"]["wp:term"][3][0]
```

Comment:
Might be worth documenting why we use this indexes.

Reply:
I agree

Reply:
I extracted all of it into "getters" functions